component-monitor/src/main.rs

```diff
@@ -0,0 +1,242 @@
+fn cleanup(target: &PathBuf, sentinel_name: &str) -> Result<()> {
+    info!(target: "files", "cleaning up {target:?}");
+    let sentinel_path = target.join(sentinel_name);
+    if sentinel_path.exists() {
+        if let Err(err) = fs::remove_file(&sentinel_path) {
+            warn!(target: "files", "failed to remove {:?}: {err}", &sentinel_path)
+        } else {
+            debug!(target: "files", "removed sentinel ({:?})", &sentinel_path)
+        }
+    }
+    for entry in fs::read_dir(target).context("failed to list target directory")? {
+        let entry = match entry {
+            Ok(e) => e,
+            Err(e) => {
+                error!(target: "files", "{e}");
+                continue;
+            }
+        };
+
+        let path = entry.path();
+        if path.is_dir() {
+            if let Err(err) = fs::remove_dir_all(&path) {
+                error!(target: "files", "failed to remove {:?}: {err}", &path)
+            } else {
+                debug!(target: "files", "removed {:?} recursively", &path)
+            }
+        } else {
+            if let Err(err) = fs::remove_file(&path) {
+                error!(target: "files", "failed to remove {:?}: {err}", &path)
+            } else {
+                debug!(target: "files", "removed {:?}", &path)
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Check if the sentinel file is current, else clean the target directory and populate again,
+/// sentinel file being last
+fn populate(source: &PathBuf, sentinel_name: &str, target: &PathBuf) -> Result<()> {
+    info!(target: "files", "populating {target:?} from {source:?} with sentinel {sentinel_name:?}");
+    let sentinel_src = source.join(sentinel_name);
+
+    if !sentinel_src.exists() {
+        info!(target: "files", "sentinel file ({:?}) not found, skipping", &sentinel_src);
+        return Ok(());
+    }
+
+    let sentinel_data = match fs::read(&sentinel_src) {
+        Ok(d) if d.trim_ascii() == b"" => {
+            error!(target: "files", "found empty sentinel file ({:?})", &sentinel_src);
+            return Ok(());
+        }
+        Ok(d) => d,
+        Err(e) => {
+            error!(target: "files", "failed to read sentinel file: {e}");
+            return Ok(());
+        }
+    };
+
+    let sentinel_tgt = target.join(sentinel_name);
+    if fs::read(&sentinel_tgt).is_ok_and(|content| content == sentinel_data) {
+        info!(target: "files", "found current sentinel, skipping");
+        return Ok(());
+    }
+
+    cleanup(target, sentinel_name)?;
+    info!(target: "files", "copying from {source:?} to {target:?}");
+    for entry in WalkDir::new(source) {
+        let entry = match entry {
+            Ok(e) => e,
+            Err(e) => {
+                error!(target: "files", "{e}");
+                continue;
+            }
+        };
+
+        let path = entry.path();
+        let relative_path = path.strip_prefix(source)?;
+        if path == source {
```

Comment:
This check is redundant since the `WalkDir` iterator already starts at `source`.

Reply:
Well, exactly, and I don't want to do anything to the source? Or do you mean that it starts _below_ the source? Don't think that was what I experienced, but will check.

Reply:
I guess I got the logic wrong; thanks for the clarification!